stl/inc/vector

```diff
@@ -2880,6 +2880,227 @@ namespace pmr {
     using vector = _STD vector<_Ty, polymorphic_allocator<_Ty>>;
 } // namespace pmr
 #endif // _HAS_CXX17
+
+#if _HAS_IF_CONSTEXPR
+// VARIABLE TEMPLATE _Is_vb_iterator
+template <class _Alloc>
+_INLINE_VAR constexpr bool _Is_vb_iterator<_Vb_const_iterator<_Alloc>> = true;
+
+template <class _Alloc>
+_INLINE_VAR constexpr bool _Is_vb_iterator<_Vb_iterator<_Alloc>> = true;
+
+template <class _InIt, class _OutIt>
+_CONSTEXPR20 _OutIt _Copy_vbool(_InIt _First, _InIt _Last, _OutIt _Dest) {
+    // copy [_First, _Last) to [_Dest, ...)
+    _Adl_verify_range(_First, _Last);
+
+    // Slow path as _First and _Dest are not aligned

```

Comment:
I am not really sure whether it is worth it actually. This was more an exercise in bit fiddling for me as I never really did any bit manipulations so far.

I am not an expert but can we do unaligned memcopy? Is that even inside the limits of the language to copy an non byte-aligned range into another non-byte aligned range?

How would that work with strict aliasing. Would we need to cast to `void*` rather than `char` to get an unaligned pointer?

I guess one could try some tricks to expand / collaps the individual bits, but that would require SSE instructions and each direction would need its own special case.

That said, in my experience vector<bool> is rather less used and I would be curious wether there are actually use cases with different ranges

Reply:
>Is that even inside the limits of the language to copy an non byte-aligned range into another non-byte aligned range?

As in 'through memcpy', no. As in 'better than extracting bit by bit', yes.

>How would that work with strict aliasing

Access as an unsigned int, no need for reinterpret here.

> I guess one could try some tricks to expand / collaps the individual bits, but that would require SSE instructions

I mean something like this pseudocode:

```
const unsigned int* data();
int main() {
const unsigned int* first = data();
size_t _Offset = 1234;
constexpr auto _Bits_per = (8*sizeof(unsigned int)); // 8 == char_bit
first += _Offset/_Bits_per;
_Offset = _Offset % _Bits_per;
unsigned int current = (first[0] >> _Offset) | first[1] << (_Bits_per - _Offset); // read 32 bools
}
```

If the target is another vector<bool> we can do a similar alignment fixup without unpacking `current`, if the target is a bool buffer that expansion would indeed be faster with AVX (since 32 bools * 8 bits == 256 bits, the AVX register size) but the same shift and mask thing can be done on a per size_t basis or similar. It won't be as fast as AVX but it will still be faster an unpacking and repacking on a bit by bit basis like before.

To be clear, I'm not saying you need to implement optimizations like that, but I am saying I believe the value of the optimization for its metaprogramming cost is questionable when the result is that `copy(v.begin(), v.end(), v2.begin())` is 10x faster than `copy(v.begin(), v.end(), v2.begin() + 1)`

Reply:
> As in 'through memcpy', no. As in 'better than extracting bit by bit', yes.

Is it too early for some warhammer references?